File: server/services/query.py
import re
from server.db.db import conn
from psycopg2.extras import DictCursor


def run(query, date_start, date_end, country=None):
    command = '''SELECT p.country as country, p.ISO as iso, p.url as paper_url, p.lang, a.url as article_url, a.publish_at, a.title_translated FROM article a
        JOIN paper p on p.uuid = a.paper_uuid
        WHERE a.publish_at >= %s
        AND DATE(a.publish_at) <= %s
        AND a.title_translated is not NULL
        ORDER BY a.publish_at desc'''

    params = (str(date_start), str(date_end))

    if country is not None:
        command += ' AND p.country = %s'
        params += (country, )

    with conn.cursor(cursor_factory=DictCursor) as c:
        c.execute(command, params)
        results = c.fetchall()

    results_matched = []
    for result in results:
        title = result['title_translated']

        print(re.search(query, title, re.IGNORECASE), query, title)
        for word in query.split(','):
            if re.search(word, title, re.IGNORECASE):
                results_matched.append(result)

    by_country = {}
    for result_matched in results_matched:
        iso = result_matched["iso"]
        if iso not in by_country:
            by_country[iso] = []
        by_country[iso].append({
            "article_url": result_matched['article_url'],
            "title": result_matched['title_translated'],
            "paper_url": result_matched['paper_url'],
            "publish_at": result_matched['publish_at'],
            "lang": result_matched['lang']
        })

    for country, articles in by_country.items():
        print('\n')
        print(country)
        print('\n')
        for article in articles:
            print(article['title'], article['article_url'])

    return by_country

File: api/query.js
import { Client } from 'pg'

export default async function handler(req, res) {
  if (req.method !== 'GET') {
    res.status(405).json({ error: 'Method not allowed' })
    return
  }

  const { query, date_start, date_end, country } = req.query
  if (!query || !date_start || !date_end) {
    res.status(400).json({ error: 'Missing params: query, date_start, date_end' })
    return
  }

  const client = new Client({ connectionString: process.env.DATABASE_URL })
  await client.connect()
  try {
    let sql = `SELECT p.country as country, p.ISO as iso, p.url as paper_url, p.lang as lang, a.url as article_url, a.publish_at, a.title_translated FROM article a
      JOIN paper p on p.uuid = a.paper_uuid
      WHERE a.publish_at >= $1
      AND DATE(a.publish_at) <= $2
      AND a.title_translated is not NULL
      ORDER BY a.publish_at desc`
    const params = [String(date_start), String(date_end)]
    if (country) {
      sql += ' AND p.country = $3'
      params.push(country)
    }
    const { rows } = await client.query(sql, params)

    // post-filter for CSV keywords (case-insensitive)
    const words = String(query).split(',').map(w => w.trim()).filter(Boolean)
    const matched = rows.filter(r => words.some(w => new RegExp(w, 'i').test(r.title_translated)))

    const byCountry = {}
    for (const r of matched) {
      const iso = r.iso
      if (!byCountry[iso]) byCountry[iso] = []
      byCountry[iso].push({
        article_url: r.article_url,
        title: r.title_translated,
        paper_url: r.paper_url,
        publish_at: r.publish_at,
        lang: r.lang
      })
    }
    res.status(200).json(byCountry)
  } catch (e) {
    console.error(e)
    res.status(500).json({ error: 'Internal error' })
  } finally {
    await client.end()
  }
}


